[SPARK-6066] Make event log format easier to parse

andrewor14 opened the pull request:
Some users have reported difficulty in parsing the new event log format. Since we embed the metadata in the beginning of the file, when we compress the event log we need to skip the metadata because we need that information to parse the log later. This means we'll end up with a partially compressed file if event logging compression is turned on. The old format looks like:

```
sparkVersion = 1.3.0
compressionCodec = org.apache.spark.io.LZFCompressionCodec
=== LOG_HEADER_END ===
// actual events, could be compressed bytes
```

The new format in this patch puts the compression codec in the log file name instead. It reformats the metadata header into JSON and includes only the Spark version. This is the first line of every event log and looks like:

```
{"Event": "SparkListenerLogStart", "Spark Version": "1.3.0"} // could be compressed
```

and the new file name looks something like:

```
app_without_compression
app_123.lzf
app_456.snappy
```

I tested this with and without compression, using different compression codecs and event logging directories. I verified that both the `Master` and the `HistoryServer` can render both compressed and uncompressed logs as before.

Commit:
Embed metadata in the event log file name instead

This makes the event logs much easier to parse than before.
As of this commit the whole file is either entirely compressed
or not compressed, but not somewhere in between.

SparkQA commented:
[Test build #28098 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28098/consoleFull) for   PR 4821 at commit [`8db5a06`](https://github.com/apache/spark/commit/8db5a06d108d8a2ddb8460e48e3509f46cc4fc2f).
 * This patch merges cleanly.

andrewor14 commented:
Tests will fail to compile. I'm fixing this in the mean time.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -110,17 +117,12 @@ private[spark] class EventLoggingListener(
-      } else {
-        None
-      }
+    val cstream = compressionCodec.map(_.compressedOutputStream(dstream)).getOrElse(dstream)
```
I know this is not part of this change, but if feels like this whole method needs a try..finally to avoid leaking an open `dstream` when some exception is thrown.

andrewor14 replied:
that's fine. I fixed this in my latest commit.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -202,11 +204,11 @@ private[spark] object EventLoggingListener extends Logging {
   val DEFAULT_LOG_DIR = "/tmp/spark-events"
 
-  private val LOG_FILE_PERMISSIONS = new FsPermission(Integer.parseInt("770", 8).toShort)
+  val EVENT_LOG_KEY = "EVENT_LOG"
```
I think the file name would be less of a mouthfull without this prefix. But not a big deal.

andrewor14 replied:
that's fine.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -217,53 +219,60 @@ private[spark] object EventLoggingListener extends Logging {
+    // e.g. EVENT_LOG_ {...} _COMPRESSION_CODEC_org.apache.spark.io.LZFCompressionCodec
+    val logName = s"${EVENT_LOG_KEY}_${sanitizedAppId}_${SPARK_VERSION_KEY}_$SPARK_VERSION" +
+      compressionCodecName.map { c => s"_${COMPRESSION_CODEC_KEY}_$c" }.getOrElse("")
+    Utils.resolveURI(logBaseDir).toString.stripSuffix("/") + "/" + logName.stripSuffix("/")
```
`stripSuffix` for `logName` seems unnecessary.

andrewor14 replied:
I was getting "/base/dir//EVENT_LOG_appId..."

vanzin replied:
But that's the `stripSuffix` on logName (first one), which is fine. The second one shouldn't be necessary.

andrewor14 replied:
oh I see what you mean. I think you're right

vanzin commented:
Looks reasonable, will wait for tests.

SparkQA commented:
[Test build #28098 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28098/consoleFull) for   PR 4821 at commit [`8db5a06`](https://github.com/apache/spark/commit/8db5a06d108d8a2ddb8460e48e3509f46cc4fc2f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28098/
Test FAILed.

Commit:
Fix tests

Commit:
Add tests for new format and file name

Commit:
Merge branch 'master' of github.com:apache/spark into event-log-format

SparkQA commented:
[Test build #28102 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28102/consoleFull) for   PR 4821 at commit [`ef69276`](https://github.com/apache/spark/commit/ef692768db319d3159ce9522d625cede3505e161).
 * This patch merges cleanly.

Commit:
Address review feedback

SparkQA commented:
[Test build #28103 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28103/consoleFull) for   PR 4821 at commit [`519e51a`](https://github.com/apache/spark/commit/519e51a958b40d193327e85b659e1df767041f55).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28102 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28102/consoleFull) for   PR 4821 at commit [`ef69276`](https://github.com/apache/spark/commit/ef692768db319d3159ce9522d625cede3505e161).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28102/
Test FAILed.

SparkQA commented:
[Test build #28103 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28103/consoleFull) for   PR 4821 at commit [`519e51a`](https://github.com/apache/spark/commit/519e51a958b40d193327e85b659e1df767041f55).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28103/
Test PASSed.

pwendell commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -217,57 +223,67 @@ private[spark] object EventLoggingListener extends Logging {
+      appId: String,
+      compressionCodecName: Option[String] = None): String = {
+    val sanitizedAppId = appId.replaceAll("[ :/]", "-").replaceAll("[${}'\"]", "_").toLowerCase
+    // e.g. EVENT_LOG_app_123_SPARK_VERSION_1.3.1
```
I don't see where the `EVENT_LOG` prefix is actually applied here.

andrewor14 replied:
you're right it's outdated

pwendell commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -217,57 +223,67 @@ private[spark] object EventLoggingListener extends Logging {
+    val sanitizedAppId = appId.replaceAll("[ :/]", "-").replaceAll("[${}'\"]", "_").toLowerCase
+    // e.g. EVENT_LOG_app_123_SPARK_VERSION_1.3.1
+    // e.g. EVENT_LOG_ {...} _COMPRESSION_CODEC_org.apache.spark.io.LZFCompressionCodec
+    val logName = s"${sanitizedAppId}_${SPARK_VERSION_KEY}_$SPARK_VERSION" +
```
Instead of totally re-writing the naming format here, can we do a more surgical change where we just add an extension for each codec? I'd use the short names here:

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/io/CompressionCodec.scala#L51

I don't see why the Spark version needs to be encoded in the name and it can't just be in the metadata header (the only thing you really need in the name is the compression codec, so that people can open the file). Also, this will keep the changes smaller in this patch.

pwendell commented on core/src/main/scala/org/apache/spark/deploy/ApplicationDescription.scala:
```diff
@@ -23,7 +23,9 @@ private[spark] class ApplicationDescription(
     val command: Command,
     var appUiUrl: String,
-    val eventLogDir: Option[String] = None)
+    val sparkVersion: String,
```
What about just passing a single option `eventLogFile` here (i.e. the full path to the log file). The master only uses the evenLogDir, sparkVersion, eventLogCodec to reconstruct the filename. Reconstructing it in two places seems a little brittle.

andrewor14 replied:
it's because at this point the `SparkContext` doesn't have an App ID yet. If we want to pass name of the file we'll need to have some place holder (e.g. `{{APP_ID}}_lzf`)

andrewor14 replied:
just adding an extra field seems like a solution that involves a smaller change, but let me know if you'd like me to change it.

pwendell commented on core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala:
```diff
@@ -99,6 +99,12 @@ case class SparkListenerExecutorRemoved(time: Long, executorId: String, reason:
+ * A special dummy event used to identify the metadata header in event logs.
+ * This is not actually posted anywhere.
+ */
+private[spark] case object SparkListenerMetadataIdentifier extends SparkListenerEvent
```
Since this is never used, why don't we instead just ignore this event type when reading from JSON (i.e. allow sparkEventFromJson to return None). It seems a bit strange to add this event but it contains nothing that's ever used.

pwendell commented:
I took a pass on this with some feedback. Overall, it would be good to really minimize the scope of the changes since this is so late in the game. There is some clean-up and renaming etc that would be best just left out of the patch.

The main thing I'm wondering is why we need this header at all. It doesn't even ever get used by our own replay - we just ignore it. It seems like it was added for the purpose of conveying the compression codec to bootstrap replaying the file, however just having an extension seems like a better, much more standard way of doing that. The only argument I see for it is that the header could be used in the future to encode things that are necessary for proper replay of the logs. However, in that case I don't see why we can't just add it later if and when those things occur.

I guess I don't see a good argument against a straw man of just not having the header. Curious to hear thoughts from @andrewor14 and @vanzin.

Commit:
Address review feedback

Things changed in this commit:
(1) No more metadata in log content
(2) No more Spark version in log file name
(3) Use short name for compression codec in log file name

Commit:
Merge branch 'master' of github.com:apache/spark into event-log-format

SparkQA commented:
[Test build #28157 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28157/consoleFull) for   PR 4821 at commit [`40e1ff6`](https://github.com/apache/spark/commit/40e1ff6398be8a55fe834e7d9c705d4ea31a520a).
 * This patch merges cleanly.

andrewor14 commented:
As of the latest commit I have removed the Spark version and the metadata. Please have a look.

pwendell commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -322,28 +322,22 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
    *
```
This doc might be better if it said "in Spark 1.2.X and earlier" instead of "in previous releases".

pwendell replied:
Actually I guess this handles multiple versions. My bad, fine to leave as is

pwendell commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -322,28 +322,22 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
    *
    * @return 2-tuple of (input stream of the events, version of Spark which wrote the log)
```
This is outdated now

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/ReplayListenerBus.scala:
```diff
@@ -39,10 +39,9 @@ private[spark] class ReplayListenerBus extends SparkListenerBus with Logging {
    * @param sourceName Filename (or other source identifier) from whence @logData is being read
    */
-  def replay(logData: InputStream, version: String, sourceName: String) {
+  def replay(logData: InputStream, sourceName: String): Unit = {
```
note that changing this signature was necessary, since the new format can no longer provide a Spark version that makes sense.

SparkQA commented:
[Test build #28158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28158/consoleFull) for   PR 4821 at commit [`59abee9`](https://github.com/apache/spark/commit/59abee953d785f8d6b4f0320eb5c5e91715c35a3).
 * This patch merges cleanly.

Commit:
Make codec an extension

New log name looks like: app123.lzf

SparkQA commented:
[Test build #28159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28159/consoleFull) for   PR 4821 at commit [`7d6aa61`](https://github.com/apache/spark/commit/7d6aa61c59a784d6f62fee94c9d7f4c6e0f501f9).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28157 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28157/consoleFull) for   PR 4821 at commit [`40e1ff6`](https://github.com/apache/spark/commit/40e1ff6398be8a55fe834e7d9c705d4ea31a520a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28157/
Test PASSed.

SparkQA commented:
[Test build #28158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28158/consoleFull) for   PR 4821 at commit [`59abee9`](https://github.com/apache/spark/commit/59abee953d785f8d6b4f0320eb5c5e91715c35a3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MatrixFactorizationModel(JavaModelWrapper, Saveable, JavaLoader):`
  * `class Saveable(object):`
  * `class Loader(object):`
  * `class JavaLoader(Loader):`
  * `        java_class = ".".join([java_package, cls.__name__])`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28158/
Test PASSed.

SparkQA commented:
[Test build #28159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28159/consoleFull) for   PR 4821 at commit [`7d6aa61`](https://github.com/apache/spark/commit/7d6aa61c59a784d6f62fee94c9d7f4c6e0f501f9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28159/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -320,30 +320,24 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
    * the code in previous releases.
    *
-   * @return 2-tuple of (input stream of the events, version of Spark which wrote the log)
+   * @return input stream that holds one JSON serialized event per line
```
nit: one JSON *record*

vanzin commented on core/src/main/scala/org/apache/spark/io/CompressionCodec.scala:
```diff
@@ -71,6 +74,21 @@ private[spark] object CompressionCodec {
+    } else {
+      shortCompressionCodecNames
+        .collect { case (k, v) if v == codecName => k }
+        .headOption
```
`collectFirst`?

andrewor14 replied:
thanks, that's exactly what I want

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -111,19 +119,17 @@ private[spark] class EventLoggingListener(
+      val cstream = compressionCodec.map(_.compressedOutputStream(dstream)).getOrElse(dstream)
+      val bstream = new BufferedOutputStream(cstream, outputBufferSize)
+      fileSystem.setPermission(path, LOG_FILE_PERMISSIONS)
+      writer = Some(new PrintWriter(bstream))
```
`new PrintWriter(bstream, "UTF-8")`

andrewor14 replied:
that signature doesn't exist. Also this was moved from L124 so it's probably fine.

vanzin replied:
http://docs.oracle.com/javase/6/docs/api/java/io/PrintWriter.html#PrintWriter(java.lang.String,%20java.lang.String)

Without that, you're writing data in the "default platform encoding".

vanzin replied:
Ah, sorry, there doesn't seem to be one for `(OutputStream, String)`. Sigh. Do you need a PrintWriter here? Can you use a regular `Writer` instead?

andrewor14 replied:
again, I'm just trying to preserve functionality here. If it's a real issue I would prefer to fix it separately, but in this patch I'm trying to minimize the scope of the changes.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -279,52 +253,17 @@ private[spark] object EventLoggingListener extends Logging {
+
+    // Compression codec is encoded as an extension, e.g. app_123.lzf
+    // Since we sanitize the app ID to not include periods, it is safe to split on it
+    val logName = log.getName.replaceAll(IN_PROGRESS, "")
```
`replaceAll` takes a regex, so you need to `Pattern.quote(IN_PROGRESS)`. In fact, it might be better to use `stripSuffix`.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -45,18 +44,26 @@ class FsHistoryProviderSuite extends FunSuite with BeforeAndAfter with Matchers
 
     // Write a new-style application log.
-    val newAppComplete = new File(testDir, "new1")
+    val newAppComplete = newLogFile("new1", inProgress = false)
```
If you're going to pass the default value explicitly, might be better to not make it an optional parameter.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -45,18 +44,26 @@ class FsHistoryProviderSuite extends FunSuite with BeforeAndAfter with Matchers
   }
 
+  /** Create a fake log file using the new log format used in Spark 1.3+ */
+  private def newLogFile(appId: String, inProgress: Boolean = false): File = {
```
Hmm... this method doesn't account for the compression codec. Maybe it should, and we should add a compressed, new-style event log to the "Parse new and old application logs" test.

andrewor14 replied:
Ok. I'm just preserving functionality here but it's not too expensive to add tests for that too.

vanzin commented:
@pwendell the header is needed because it contains potentially useful information for the code parsing the logs. For example, now it contains the Spark version, which might be needed to tell the parsing code which properties to expect in the logs.

The original version of the change (the one that got rid of the directories and used a single file) encoded all metadata in the file name. The feedback was that it was ugly (long, cryptic file names) and brittle, since if you change the file name, you lose that information. I agree with that and thus the header was born.

Now we're back to encoding metadata in the file name. A simple extension is not to bad, though, espcially since you can probably figure out the compression codec by looking at the first few bytes of the file. But the header still provides useful information.

So I'm a little worried that the latest patch removes the metadata completely. Especially since it's common for the first event of the log to *not* be the one that contains the spark version (`SparkListenerEnvironmentUpdate`?), and instead be `SparkListenerBlockManagerAdded`.

pwendell commented:
@vanzin so just to be clear - you are anticipating a future case where we need to read the version to correctly parse the logs? Is that the argument for it? I am only making a simple observation that here we write something and never read it.

vanzin commented:
@pwendell yes, that's what I'm trying to say. Even today, logs generated by Spark 1.0, 1.1 and 1.2 (and soon 1.3) are not exactly the same - events have different properties. The metadata allows the parsing code to know what to expect.

(That is not a problem internally in Spark because the code handles all added/changed event properties using optionals and things like that, but there's no way to know whether we'll make a change in the future that prevents code like that from working without knowing the version of the data.)

Commit:
Address review feedback

SparkQA commented:
[Test build #28180 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28180/consoleFull) for   PR 4821 at commit [`7f537cd`](https://github.com/apache/spark/commit/7f537cda6e68f922889f5b95e2869ec770f038a8).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28180 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28180/consoleFull) for   PR 4821 at commit [`7f537cd`](https://github.com/apache/spark/commit/7f537cda6e68f922889f5b95e2869ec770f038a8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28180/
Test FAILed.

pwendell commented:
@vanzin okay then, let's add back a JSON metadata header to the file.

Commit:
Add back metadata with Spark version

andrewor14 commented:
By the way I gained first-hand experience of the flakiness of `FsHistoryProviderSuite` from this patch. The tests passed locally but failed here, and I suspect that it has something to do with the modification time not varying enough between the file writes, though this is something we should fix separately.

SparkQA commented:
[Test build #28182 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28182/consoleFull) for   PR 4821 at commit [`654883d`](https://github.com/apache/spark/commit/654883dfbd65de162455601c69108dab4e354f7d).
 * This patch merges cleanly.

vanzin commented:
>  flakiness of FsHistoryProviderSuite

Do you have any theories for why this would be related to the log's modification time? The only place where it's used in FsHistoryProvider is for filtering entries in `checkForLogs`, and it's already a conservative check (it uses `>=` so that it picks up any new logs that show up with the same mod time as the last check, at the expense of re-parsing logs that haven't changed).

SparkQA commented:
[Test build #28182 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28182/consoleFull) for   PR 4821 at commit [`654883d`](https://github.com/apache/spark/commit/654883dfbd65de162455601c69108dab4e354f7d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28182/
Test FAILed.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -89,16 +105,18 @@ class FsHistoryProviderSuite extends FunSuite with BeforeAndAfter with Matchers
     list(0) should be (ApplicationHistoryInfo(newAppComplete.getName(), "new-app-complete", 1L, 4L,
       newAppComplete.lastModified(), "test", true))
-    list(1) should be (ApplicationHistoryInfo(oldAppComplete.getName(), "old-app-complete", 2L, 3L,
+    list(1) should be (ApplicationHistoryInfo(newAppCompressedComplete.getName(),
```
I think your problem is that entires `1` and `2` have both the same start and end time, so their sort order is non-deterministic. Also, `newAppCompressedComplete` has end time `4` in the code a few lines above. So it seems your test code has a couple of issues, this is not really flakiness.

andrewor14 replied:
yeah, I just realized this independently. I fixed this in my latest commit.

andrewor14 replied:
however that still doesn't explain why it passes locally but fails remotely.

vanzin replied:
It could be that you're using a different filesystem than then jenkins machine, and both return things in different order.

Commit:
Fix test

SparkQA commented:
[Test build #28198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28198/consoleFull) for   PR 4821 at commit [`fdae14c`](https://github.com/apache/spark/commit/fdae14c86f660047b50020b40a782b842182336c).
 * This patch merges cleanly.

pwendell commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -574,6 +583,11 @@ private[spark] object JsonProtocol {
   }
 
+  def logStartFromJson(json: JValue): SparkListenerLogStart = {
+    val version = (json \ "Spark Version").extract[String]
```
minor - but maybe call this `sparkVersion` so it's clear that this isn't a version for the logging format (since that doesn't have its own versioning).

andrewor14 replied:
but then it's not consistent with other similar JSONs...

andrewor14 replied:
(edit): misunderstood what you meant. I'll rename the variable NBD

vanzin commented:
LGTM.

SparkQA commented:
[Test build #28201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28201/consoleFull) for   PR 4821 at commit [`8511141`](https://github.com/apache/spark/commit/8511141f6b769655acc2b0c1de33a8401c8f7133).
 * This patch merges cleanly.

pwendell commented:
Okay I took a close look through this and it LGTM

SparkQA commented:
[Test build #28198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28198/consoleFull) for   PR 4821 at commit [`fdae14c`](https://github.com/apache/spark/commit/fdae14c86f660047b50020b40a782b842182336c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28198/
Test PASSed.

SparkQA commented:
[Test build #28201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28201/consoleFull) for   PR 4821 at commit [`8511141`](https://github.com/apache/spark/commit/8511141f6b769655acc2b0c1de33a8401c8f7133).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28201/
Test PASSed.
